### PR TITLE
bump python version to 3.9

### PR DESCRIPTION
**Pull request by e-davidson**
*Issue #, if available:*
python 3.8 is deprecated
*Description of changes:*

Update to python 3.9

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docker/1.2-1/base/Dockerfile.cpu [hunk 1]
```diff
@@ -89,10 +89,10 @@ RUN echo "conda ${CONDA_PKG_VERSION}" >> /miniconda3/conda-meta/pinned && \
     conda config --system --set auto_update_conda false && \
     conda config --system --set show_channel_urls true && \
     echo "python ${PYTHON_VERSION}.*" >> /miniconda3/conda-meta/pinned && \
-    conda install -c conda-forge python=${PYTHON_VERSION} && \
-    conda install conda=${CONDA_PKG_VERSION} && \
+    conda install -c conda-forge python=${PYTHON_VERSION} --solver classic && \
+    conda install conda=${CONDA_PKG_VERSION} --solver classic && \
```
Comment (malav-shastri) [inline docker/1.2-1/base/Dockerfile.cpu:93]: just curious, what value does --solver classic add?

**Comment by malav-shastri** [comment]
lgtm overall

File: docker/1.2-1/base/Dockerfile.cpu (hunk 1 above)
Comment (e-davidson) [reply docker/1.2-1/base/Dockerfile.cpu:93]: The install command failed without it. The error message said to do that.

**Review by malav-shastri — APPROVED** [review]

**Comment by w1l50n** [comment]
I'm late to the party. does python 3.9 work with sagemaker-containers==2.8.6.post2 (deprecated 4yrs ago)? 

https://github.com/aws/sagemaker-scikit-learn-container/blob/master/requirements.txt#L14
https://github.com/aws/sagemaker-containers (deprecated)

does anyone have luck migrate this container to only use [sagemaker-training-toolkit](https://github.com/aws/sagemaker-training-toolkit) and [sagemaker-inference-toolkit](https://github.com/aws/sagemaker-inference-toolkit)?

**Comment by w1l50n** [comment]
I just pull build the container for this update and it works. thanks

File: docker/1.2-1/base/Dockerfile.cpu (hunk 1 above)
Comment (w1l50n) [reply docker/1.2-1/base/Dockerfile.cpu:93]: newer conda=>23.10 have libmamba as default solver
https://conda.github.io/conda-libmamba-solver/user-guide/
https://conda.github.io/conda-libmamba-solver/user-guide/libmamba-vs-classic/

my error was 
```
35.21 Error while loading conda entry point: conda-libmamba-solver (libarchive.so.20: cannot open shared object file: No such file or directory)
35.31
35.31 CondaValueError: You have chosen a non-default solver backend (libmamba) but it was not recognized. Choose one of: classic
35.31
``` 

you could either use `--solver classic` or forcing all channels to use conda-forge
```shell
--- a/docker/1.2-1/base/Dockerfile.cpu
+++ b/docker/1.2-1/base/Dockerfile.cpu
@@ -88,11 +88,13 @@ RUN echo "conda ${CONDA_PKG_VERSION}" >> /miniconda3/conda-meta/pinned && \
     # Conda configuration see https://conda.io/projects/conda/en/latest/configuration.html
     conda config --system --set auto_update_conda false && \
     conda config --system --set show_channel_urls true && \
+    conda config --add channels conda-forge && \
     echo "python ${PYTHON_VERSION}.*" >> /miniconda3/conda-meta/pinned && \
-    conda install -c conda-forge python=${PYTHON_VERSION} --solver classic && \
-    conda install conda=${CONDA_PKG_VERSION} --solver classic && \
+    conda update --all --override-channels -c conda-forge -n base && \
+    conda install -c conda-forge python=${PYTHON_VERSION} && \
+    conda install conda=${CONDA_PKG_VERSION} && \
     conda update -y conda && \
-    conda install -c conda-forge pyarrow=${PYARROW_VERSION} --solver classic && \
+    conda install -c conda-forge pyarrow=${PYARROW_VERSION} && \
     conda install pip --force-reinstall && \
     python3 -m pip install --upgrade pip && \
     python3 -m pip install pyOpenSSL==23.1.0 && \
```

**Comment by maxiuboldi** [comment]
Hi, at my work we are starting to have failures in production and we believe it is due to this change. Do you think it is a good practice to modify the Python version of the container without versioning the container? We have additional dependencies in the `framework_version='1.2-1'` that were tied to Python 3.8 and today we are finding processes failing because those dependencies are not compatible with version 3.9.

**Comment by w1l50n** [comment]
> Hi, at my work we are starting to have failures in production and we believe it is due to this change. Do you think it is a good practice to modify the Python version of the container without versioning the container? We have additional dependencies in the `framework_version='1.2-1'` that were tied to Python 3.8 and today we are finding processes failing because those dependencies are not compatible with version 3.9.

@maxiuboldi do you mind to share what dependencies doesn't work with Python 3.9?

**Comment by maxiuboldi** [comment]
Hi, thanks for the reply @w1l50n. I don't understand why you would be interested in knowing my dependencies. My question is related to whether you think it's a good practice to modify the Python version of the framework in this way. We have jobs running that were developed using Python 3.8 and now we see that for exactly the same conditions that we had in production yesterday, you modified the environment for us. I think it's perfect that you update the versions of the framework, in fact it seems more than necessary, but what you can't do, in my opinion, is to do it without versioning the image. Yesterday my job ran fine with Python 3.8 and today it no longer runs with Python 3.9, it doesn't matter how I defined the job or the dependencies I have, you can't do that, that's what versions and tags are for and, unlike DLC images, in this framework we can't choose the Python version to use from the SDK, only the framework version.

**Comment by w1l50n** [comment]
Hi @maxiuboldi, 
I'm interested because I may ran into same failure. so it will be good to know which depending library failed.
I hear your frustration. I agree that the framework version should be updated too. Did AWS already update the docker container image on their ECR that build from this update? If that is the case, I assume there will be many unhappy folks.

What I did was to build the docker image and upload to my own ECR. Name it something like **sklearn:1.2-1-cpu-py39**, then use the **image_uri** for training and deployment instead of **framework_version** when calling the sagemaker.

**Comment by maxiuboldi** [comment]
Sorry @w1l50n, this is a serious problem for me. Yes, I understand AWS updated the image today because this started happening today at my work. And yes, your proposal may be an alternative, but we have several models running under this image in production and we can't anticipate all the ones that may fail because each one may or may not have additional dependencies that could potentially conflict. Is there no way to undo this?

I have no problem sharing with you some failed requirement.txt later, but right now it seems more urgent to me to find a solution that doesn't involve me being aware of potential failures and looking for ways to fix them, there are processes that are critical for the business and this is not good at all.

**Comment by malav-shastri** [comment]
Hi @maxiuboldi apologies for this inconvenience. We hear your concern and will make changes to our processes moving forward. Python 3.8 reached EOL and it was important for us to update the version as soon as possible. Without wasting your time I would suggest you to pin the image to this image digest
sha256:e09bbb7686077a1db23d316b699020a786a6e1636b2b89384be9651368c40f95
This is the image which has python 3.8

However this is not a long term solution at all and it's my responsibility to keep you informed about the side effects of this, you'll not receive any security vulnerability patches from our side as long as you're pinning the image sha. And as python 3.8 is EOL I would recommend to update your prod environment and migrate to python 3.9 specific dependencies as long as possible.

**Comment by maxiuboldi** [comment]
Hi @malav-shastri, thanks for your reply. I am aware of the drawbacks of continuing to use Python 3.8 (and I actually wonder why this framework, being one of the most used, is so outdated in SageMaker) but that does not imply that you change the environments in this way. I insist, that is what versions are for. I fully understand that you warn me and that you no longer support this version, but I do not agree that you change it in this way. Also, until yesterday, 3.8 was the last version supported by the framework.

In any case, I would appreciate it if you could review the circuit and keep the framework updated, so that we do not see ourselves obliged on our side to maintain obsolete code and versions even in new developments.

**Comment by maxiuboldi** [comment]
@w1l50n here I share with you a test `requirements.txt` that we used to reproduce the problem. In this particular case it seems to be an incompatibility with the `scikit-plot` dependency, which depends on `matplotlib` and `numpy`.

image

[requirements.txt](https://github.com/user-attachments/files/17441844/requirements.txt)

**Comment by w1l50n** [comment]
@malav-shastri you suggested to pin the image digest to sha256@e09bbb7686077a1db23d316b699020a786a6e1636b2b89384be9651368c40f95 did you mean I should pull the image to my own ECR and use it from there or I could use it as the image_uri param in the [sagemaker python sdk](https://sagemaker.readthedocs.io/en/stable/frameworks/sklearn/sagemaker.sklearn.html)? e.g. 
```python
sagemaker.sklearn.estimator.SKLearn(
    ...,
    image_uri="257758044811.dkr.ecr.us-east-2.amazonaws.com/sagemaker-scikit- 
    learn:@sha256@e09bbb7686077a1db23d316b699020a786a6e1636b2b89384be9651368c40f95",
    ...
)
``` ?

I also want to discuss about future Python version update (e.g. 3.10, 3.11, etc). I tried to patch this repo in a minimal way to run on Python 3.10 by:
1. remove the deprecated [sagemaker-containers](https://github.com/aws/sagemaker-containers), and use whatever available from [sagemaker-training-toolkit](https://github.com/aws/sagemaker-training-toolkit) and [sagemaker-inference-toolkit](https://github.com/aws/sagemaker-inference-toolkit)
2. patch whatever needed so I can use the container to train and deploy
However, it was not a trivial as I thought. And I saw this pull request literally on the day I was about to flip the table 😄. I probably should have tried 3.9 first, but who don't like the latest and greatest.

Do you happen to know if AWS is planning to release a sagemaker-sklearn-training-toolkit and sagemaker-sklearn-inference-toolkit as a replacement? If not what's your suggestion for folks wanting to use newer Python version like 3.11? I hope I'm not the only one who want to use the latest and greatest. Or it is very trivial to build the scikit-learn container from scratch using [sagemaker-training-toolkit](https://github.com/aws/sagemaker-training-toolkit) and [sagemaker-inference-toolkit](https://github.com/aws/sagemaker-inference-toolkit), so no one is bother to ask?

**Comment by w1l50n** [comment]
@maxiuboldi thanks for sharing 
looks like numpy._core wasn't add util 1.26. may be there some unpinned dependency causing the dependency spaghetti 😨

**Comment by maxiuboldi** [comment]
> @maxiuboldi thanks for sharing looks like numpy._core wasn't add util 1.26. may be there some unpinned dependency causing the dependency spaghetti 😨

That requirements was a test in development to reproduce the error, it is not the requirements we have in production. In production the versions are fixed, but the error is the same. Sorry if I did not express myself well, my English is not very good 😅

This is production and just failed in inference:

> lightgbm==4.1.0
> pandas==2.0.3
> numpy==1.24.4
> imbalanced-learn==0.11.0
> category-encoders==2.6.2
> feature-engine==1.6.2
> scikit-plot==0.3.7
> str2bool==1.1

image

In Python 3.8 it has been working without problems for months...

**Comment by w1l50n** [comment]
@maxiuboldi not sure what your serving entry point script looks like, but you could try to expend the stacktrace to see if there is any hint on which module is trying to call `numpy._core.multiarray`, track down that library version, and hopefully find the version plays well with the numpy ==1.24.4 and add that to the requirements file
378043025-c5f9d4b5-b9d2-44b3-b885-e55e5e6ec7ea

**Comment by maxiuboldi** [comment]
Yes, I can fix the problem, adapt the versions to Python 3.9, modify whatever is necessary, but that is not the point. The point is that this change should not have been done this way. This is not the way to upgrade, without versioning. Nor is it the way to deprecate a version that is not supported. We develop and test the code in Python 3.8 and it is not right that the environment is changed in this way and in production. We have a defined circuit for changes in production, it is not as easy to make changes as if it were in development and it is also affecting the business and the users.

In any case, I appreciate your help.

**Comment by malav-shastri** [comment]
> @malav-shastri you suggested to pin the image digest to sha256@e09bbb7686077a1db23d316b699020a786a6e1636b2b89384be9651368c40f95 did you mean I should pull the image to my own ECR and use it from there or I could use it as the image_uri param in the [sagemaker python sdk](https://sagemaker.readthedocs.io/en/stable/frameworks/sklearn/sagemaker.sklearn.html)? e.g.
> 
> ```python
> sagemaker.sklearn.estimator.SKLearn(
>     ...,
>     image_uri="257758044811.dkr.ecr.us-east-2.amazonaws.com/sagemaker-scikit- 
>     learn:@sha256@e09bbb7686077a1db23d316b699020a786a6e1636b2b89384be9651368c40f95",
>     ...
> )
> ```
> 
> I also want to discuss about future Python version update (e.g. 3.10, 3.11, etc). I tried to patch this repo in a minimal way to run on Python 3.10 by:
> 
> 1. remove the deprecated [sagemaker-containers](https://github.com/aws/sagemaker-containers), and use whatever available from [sagemaker-training-toolkit](https://github.com/aws/sagemaker-training-toolkit) and [sagemaker-inference-toolkit](https://github.com/aws/sagemaker-inference-toolkit)
> 2. patch whatever needed so I can use the container to train and deploy
>    However, it was not a trivial as I thought. And I saw this pull request literally on the day I was about to flip the table 😄. I probably should have tried 3.9 first, but who don't like the latest and greatest.
> 
> Do you happen to know if AWS is planning to release a sagemaker-sklearn-training-toolkit and sagemaker-sklearn-inference-toolkit as a replacement? If not what's your suggestion for folks wanting to use newer Python version like 3.11? I hope I'm not the only one who want to use the latest and greatest. Or it is very trivial to build the scikit-learn container from scratch using [sagemaker-training-toolkit](https://github.com/aws/sagemaker-training-toolkit) and [sagemaker-inference-toolkit](https://github.com/aws/sagemaker-inference-toolkit), so no one is bother to ask?

@w1l50n We really appreciate you taking a keen interest in this container. for the questions you asked,
- you could use pinned image uri as the image_uri param in the [sagemaker python sdk](https://sagemaker.readthedocs.io/en/stable/frameworks/sklearn/sagemaker.sklearn.html) 
- We are aware about the deprecated sagemaker-containers and planning to switch to sagemaker-training-toolkit and sagemaker-inference-toolkit to unblock latest python version upgrades as well as new SKlearn major version release. We also need to make sure that this container remains compatible with other sagemaker functionalities and there are some low level serving stack related decisions which needs to be made before actually getting rid of sagemaker-containers completely. Actually the container currently has a partial dependency on those new toolkits already, multi model endpoint feature already uses it. 
- Apart from all these, this container open source and I would like to take this opportunity to invite valuable community members like yourself to help us maintain this. We can make sure to review your suggestions and help merging your PRs on a priority basis.

**Comment by w1l50n** [comment]
@malav-shastri Thanks for the update. It is great to hear that AWS is planning to update the scikit learn container. I would like to help out to the best of my ability. what is good communication channel for about that? it is possible to add "Issues" or "Discussions" like the [sagemaker-python-sdk](https://github.com/aws/sagemaker-python-sdk) repo?
Screenshot 2024-10-21 at 9 21 57 AM

**Comment by malav-shastri** [comment]
> @malav-shastri Thanks for the update. It is great to hear that AWS is planning to update the scikit learn container. I would like to help out to the best of my ability. what is good communication channel for about that? it is possible to add "Issues" or "Discussions" like the [sagemaker-python-sdk](https://github.com/aws/sagemaker-python-sdk) repo? Screenshot 2024-10-21 at 9 21 57 AM

@w1l50n You can add issues or discussion in SKLearn repo itself. Do you see that option here?

**Comment by malav-shastri** [comment]
> @malav-shastri I imagine the Python 3.9 update becomes a breaking change for some customers. I'm wondering if AWS is going to:
> 
> 1. keep the Scikit-Learn 1.2-1 container with Python 3.9 ?
> 2. rollback the Scikit-Learn 1.2-1 container to build with Python 3.8, and build a new one (e.g. 1.2-1.1) with Python 3.9 ?

In my understanding the next focus for us should be to invest in migrating to training-toolkit and inference-toolkit libs to unblock next set of updates. Customers can easily recover from this by pinning the sha and we'll be helping them with that.

**Comment by w1l50n** [comment]
Hi @malav-shastri,

-  Could you please share the link to the **SKLearn** repo you were talking about? if you are talking about this repo [sagemaker-scikit-learn-container](https://github.com/aws/sagemaker-scikit-learn-container). I don't see the discussion option. Screenshot 2024-10-22 at 1 40 32 PM

- Thanks for sharing the intend on Scikit-Learn 1.2-1 container with Python 3.9

**Comment by malav-shastri** [comment]
> Hi @malav-shastri,
> 
> * Could you please share the link to the **SKLearn** repo you were talking about? if you are talking about this repo [sagemaker-scikit-learn-container](https://github.com/aws/sagemaker-scikit-learn-container). I don't see the discussion option. Screenshot 2024-10-22 at 1 40 32 PM
> * Thanks for sharing the intend on Scikit-Learn 1.2-1 container with Python 3.9

@w1l50n apologies for delayed response. I see, let me take a followup on enabling Issues tab here in this repo.

**Comment by malav-shastri** [comment]
> Hi @malav-shastri,
> 
> * Could you please share the link to the **SKLearn** repo you were talking about? if you are talking about this repo [sagemaker-scikit-learn-container](https://github.com/aws/sagemaker-scikit-learn-container). I don't see the discussion option. Screenshot 2024-10-22 at 1 40 32 PM
> * Thanks for sharing the intend on Scikit-Learn 1.2-1 container with Python 3.9

I have enabled the issues section for this repo so that we can collaborate more effectively. Thanks

**Comment by malav-shastri** [comment]
> Hi @malav-shastri,
> 
> * Could you please share the link to the **SKLearn** repo you were talking about? if you are talking about this repo [sagemaker-scikit-learn-container](https://github.com/aws/sagemaker-scikit-learn-container). I don't see the discussion option. Screenshot 2024-10-22 at 1 40 32 PM
> * Thanks for sharing the intend on Scikit-Learn 1.2-1 container with Python 3.9

@w1l50n I have enabled the issues section for this repo so that we can collaborate more effectively. Thanks

**Pull request merged**